vembrane/__init__.py

```diff
@@ -111,77 +180,187 @@ def get_annotation_keys(header: VariantHeader, ann_key: str) -> List[str]:
+        avaible_vcf_fields = {
+            "QUAL",
+            "FILTER",
+            "ID",
+            "CHROM",
+            "POS",
+            "REF",
+            "ALT",
+            "SAMPLES",
+        } & names
+        available_symbols = globals_whitelist.keys() & names
+        self.globals_whitelist = {
+            key: globals_whitelist[key] for key in available_symbols
+        }
+
+        ann_field_name = expression.annotation_key()
+        annotation_keys = get_annotation_keys(vcf_header, ann_field_name)
+        # For the annotation keys, we have to keep track of their indices,
+        # since the annotation values are a string joined by (usually) '|',
+        # i.e. when splitting at '|', extract only the values at certain indices.
+        available_annotation_keys = [
+            (i, k) for i, k in enumerate(annotation_keys) if k in names
+        ]
+
+        self.idx = -1  # Consider changing this to None
+
+        # For INFO, ANN and FORMAT, build empty containers
+        # (with keys appearing in the expression set to some default)
+        self.info = Info(
+            record_idx=self.idx,
+            info_dict={
+                info_key: {}
+                for info_key in available_info_fields
+                if info_key != ann_field_name
+            },
+        )
+        self.annotation = Annotation(
+            record_idx=self.idx,
+            annotation_data={field: None for (_, field) in available_annotation_keys},
+            available_keys=available_annotation_keys,
+        )
+        self.sample_names = available_sample_names
+        self.formats = Format(
+            record_idx=self.idx,
+            sample_formats={
+                sample: Sample(record_idx=self.idx, sample=sample, format_data={})
+                for sample in self.sample_names
+            },
+            format_keys=available_format_keys,
+        )
+
+        # These are the usual VCF columns (plus "SAMPLES")…
+        self.field_lookup = {
+            "CHROM": lambda record: record.chrom,
+            "POS": lambda record: record.pos,
+            "ID": lambda record: record.id,
+            "QUAL": lambda record: type_info(record.qual),
+            "FILTER": lambda record: record.filter,
+            "REF": lambda record: record.alleles[0],
+            "ALT": lambda record: record.alleles[1:],
+            "SAMPLES": lambda record: list(record.samples),
+        }
+
+        # … but we don't need all of them, so remove unneeded fields:
+        for field in self.field_lookup.keys() - set(avaible_vcf_fields):
+            self.field_lookup.pop(field, None)
+
+        # Build the env dict used for `eval`
+        self.env = {
+            "INFO": self.info,
+            "FORMAT": self.formats,
+            ann_field_name: self.annotation,
+        }
+
+        self.globals = {}
+        self.func = eval(f"lambda: {expression.raw_expression}", self.globals, {})
```

Comment:
if you run `eval` here, does it actually keep the *reference* of `self.globals`? I mean, since the tests don't fail, it must be that way :D

Reply:
In any case, I obviously wasn't aware of the precise `eval` semantics and thought we had to call that everytime something changes, even though the expression (and hence the "compilation") itself does not change.
Very neat trick ;)

Just gave this a try, with your change running time went down to 5s! Awesome!

Reply:
Ah, should've made a comment!
Yep, I use `globals` as a closure here. Hence, we should never reassign `self.globals = ...` and also inject `self.env` into `self.globals` and we're set :).

Reply:
> and also inject `self.env` into `self.globals` and we're set :).

but you already do that in `evaluate`: `self.globals.update(self.env)`, right?

Reply:
Exactly, I was referring to that line ;)